pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -396,96 +396,115 @@ static PyObject *get_item(PyObject *obj, Py_ssize_t i) {
     return ret;
 }
 
-static void *PyBytesToUTF8(JSOBJ _obj, JSONTypeContext *tc, void *outValue,
-                           size_t *_outLen) {
+static char *PyBytesToUTF8(JSOBJ _obj, JSONTypeContext *tc, size_t *_outLen) {
     PyObject *obj = (PyObject *)_obj;
     *_outLen = PyBytes_GET_SIZE(obj);
     return PyBytes_AS_STRING(obj);
 }
 
-static void *PyUnicodeToUTF8(JSOBJ _obj, JSONTypeContext *tc, void *outValue,
-                             size_t *_outLen) {
-  return PyUnicode_AsUTF8AndSize(_obj, _outLen);
+static char *PyUnicodeToUTF8(JSOBJ _obj, JSONTypeContext *tc, size_t *_outLen) {
+    return (char *)PyUnicode_AsUTF8AndSize(_obj, _outLen);
 }
 
-static void *PandasDateTimeStructToJSON(npy_datetimestruct *dts,
-                                        JSONTypeContext *tc, void *outValue,
-                                        size_t *_outLen) {
-    NPY_DATETIMEUNIT base = ((PyObjectEncoder *)tc->encoder)->datetimeUnit;
+/* returns a char* and mutates the pointer to *len */
+static char *NpyDateTimeToIso(JSOBJ unused, JSONTypeContext *tc, size_t *len) {
+    npy_datetimestruct dts;
+    int ret_code;
+    int64_t longVal = GET_TC(tc)->longValue;
 
-    if (((PyObjectEncoder *)tc->encoder)->datetimeIso) {
-        PRINTMARK();
-        *_outLen = (size_t)get_datetime_iso_8601_strlen(0, base);
-        GET_TC(tc)->cStr = PyObject_Malloc(sizeof(char) * (*_outLen));
-        if (!GET_TC(tc)->cStr) {
-            PyErr_NoMemory();
-            ((JSONObjectEncoder *)tc->encoder)->errorMsg = "";
-            return NULL;
-        }
+    pandas_datetime_to_datetimestruct(longVal, NPY_FR_ns, &dts);
 
-        if (!make_iso_8601_datetime(dts, GET_TC(tc)->cStr, *_outLen, base)) {
-            PRINTMARK();
-            *_outLen = strlen(GET_TC(tc)->cStr);
-            return GET_TC(tc)->cStr;
-        } else {
-            PRINTMARK();
-            PyErr_SetString(PyExc_ValueError,
-                            "Could not convert datetime value to string");
-            ((JSONObjectEncoder *)tc->encoder)->errorMsg = "";
-            PyObject_Free(GET_TC(tc)->cStr);
-            return NULL;
-        }
-    } else {
-        PRINTMARK();
-        *((JSINT64 *)outValue) = npy_datetimestruct_to_datetime(base, dts);
+    NPY_DATETIMEUNIT base = ((PyObjectEncoder *)tc->encoder)->datetimeUnit;
+    *len = (size_t)get_datetime_iso_8601_strlen(0, base);
+    char *result = PyObject_Malloc(*len);
+
+    if (result == NULL) {
```

Comment:
how would we get here?  if `len == 0`?

Reply:
This would happen if malloc fails

https://stackoverflow.com/a/12434865/621736

Reply:
So failing to address it would mean getting something like a segfault?  the PyErr_NoMemory just below raises back in py-space?

Reply:
I think it would segfault if you tried to access the memory if malloc failed to allocate it. `PyErr_NoMemory` would just set the global error indicator so when the extension exits that should raise back in the Python space (assuming nothing else clears it out)

FWIW the error handling here is copy / paste from here:

https://github.com/pandas-dev/pandas/blob/98cb43267eec34203e2a169cbf7c79bed7e9a9c3/pandas/_libs/src/ujson/python/objToJSON.c#L422

There are similar checks elsewhere throughout the module (not always consistent) so maybe worth unifying in a follow up